File: src/schur-weyl/symmetric_group.py
from itertools import permutations
from typing import Generator
from collections import Counter, defaultdict
from math import factorial


def permutation_identity(n) -> tuple[int, ...]:
    """The identity permutation on {0, ..., n-1}.

    Examples:
        >>> permutation_identity(4)
        (0, 1, 2, 3)
    """
    return tuple(range(n))


def permutation_compose(pi: tuple[int, ...], sigma: tuple[int, ...]) -> tuple[int, ...]:
    """Multiply two permutation (pi * sigma)(i) = pi(sigma(i))

    Examples:
        >>> permutation_compose((1, 0, 2), (0, 2, 1))
        (1, 2, 0)
        >>> permutation_compose((0, 1, 2), (2, 0, 1))
        (2, 0, 1)
    """
    return tuple(pi[sigma[i]] for i in range(len(pi)))


def permutation_inverse(sigma: tuple[int, ...]) -> tuple[int, ...]:
    """The inverse permutation.

    Examples:
        >>> permutation_inverse((1, 2, 0))
        (2, 0, 1)
        >>> permutation_inverse((0, 1, 2))
        (0, 1, 2)
    """
    result = [0] * len(sigma)
    for i, s in enumerate(sigma):
        result[s] = i
    return tuple(result)


def all_permutations(n: int) -> Generator[tuple[int, ...]]:
    """All permutations of Sn

    Args:
        n (int): n

    Yields:
        Generator[tuple[int, ...]]: the permutations
    """
    for p in permutations(range(n)):
        yield p


def all_permutations_by_cycle_type(n: int) -> dict[tuple[int, ...], list[tuple[int, ...]]]:
    """All permutations of Sn assigned to their cycle type

    Args:
        n (int): n

    Yields:
        dict[tuple[int, ...], list[tuple[int, ...]]]: the cycle type and its permutations
    """
    cycle_and_perm: dict[tuple[int, ...], list[tuple[int, ...]]] = defaultdict(list)
    for p in permutations(range(n)):
        cycle_and_perm[permutation_cycle_type(p)].append(p)
    return cycle_and_perm


def permutation_cycle_type(perm: tuple[int, ...]) -> tuple[int, ...]:
    """Give the cycle type of a permutation. The disjoint sets of perms

    Args:
        perm (tuple[int, ...]): The permutation

    Returns:
        tuple[int, ...]: Cycle type

    Examples:
        >>> permutation_cycle_type((1, 2, 0, 4, 3))
        (3, 2)
        >>> permutation_cycle_type((0, 1, 2))
        (1, 1, 1)
        >>> permutation_cycle_type((1, 0, 2))
        (2, 1)
    """

    seen: list[bool] = [False] * len(perm)
    cycles: list[int] = []
    for i in range(len(perm)):
        if seen[i]:
            continue
        seen[i] = True

        count = 1
        j = perm[i]
        while not seen[j]:
            seen[j] = True
            j = perm[j]
            count += 1
        cycles.append(count)
    return tuple(sorted(cycles, reverse=True))


def permutation_conjugacy_class_size(cycle_type: tuple[int, ...]) -> int:
    """The number of permutations of the same cycle type

    Args:
        cycle_type (tuple[int, ...]): The cycle type

    Returns:
        int: The number of permutations of the same cycle type

    Examples:
        >>> permutation_conjugacy_class_size((1,1,1,1))
        1
        >>> permutation_conjugacy_class_size((2,1,1))
        12
    """
    counts = Counter(cycle_type)
    z = 1
    for cycle_len, cycle_count in counts.items():
        z = cycle_len**cycle_count * factorial(cycle_count)
    return factorial(sum(cycle_type)) // z

File: docs/src/schur-weyl/symmetricGroup.ts
/**
 * Ported from `src/schur_weyl/symmetric_group.py`.
 *
 * A permutation of {0, ..., n-1} is represented as its 0-indexed one-line
 * array, e.g. `[1, 2, 0]` sends 0 -> 1, 1 -> 2, 2 -> 0.
 */

import { factorial } from './utils';

export type Permutation = number[];

/**
 * The identity permutation on {0, ..., n-1}.
 *
 * @example
 * permutationIdentity(4) // [0, 1, 2, 3]
 */
export function permutationIdentity(n: number): Permutation {
  return Array.from({ length: n }, (_, i) => i);
}

/**
 * Multiply two permutations: (pi * sigma)(i) = pi(sigma(i)).
 *
 * @example
 * permutationCompose([1, 0, 2], [0, 2, 1]) // [1, 2, 0]
 * permutationCompose([0, 1, 2], [2, 0, 1]) // [2, 0, 1]
 */
export function permutationCompose(pi: Permutation, sigma: Permutation): Permutation {
  return sigma.map((_, i) => pi[sigma[i]]);
}

/**
 * The inverse permutation.
 *
 * @example
 * permutationInverse([1, 2, 0]) // [2, 0, 1]
 * permutationInverse([0, 1, 2]) // [0, 1, 2]
 */
export function permutationInverse(sigma: Permutation): Permutation {
  const result = new Array<number>(sigma.length).fill(0);
  sigma.forEach((s, i) => {
    result[s] = i;
  });
  return result;
}

/** All permutations of {0, ..., n-1}. */
export function* allPermutations(n: number): Generator<Permutation, void, void> {
  yield* permuteRange(Array.from({ length: n }, (_, i) => i));
}

function* permuteRange(arr: number[]): Generator<Permutation, void, void> {
  if (arr.length <= 1) {
    yield [...arr];
    return;
  }
  for (let i = 0; i < arr.length; i++) {
    const rest = [...arr.slice(0, i), ...arr.slice(i + 1)];
    for (const perm of permuteRange(rest)) {
      yield [arr[i], ...perm];
    }
  }
}

/**
 * Give the cycle type of a permutation: the sizes of its disjoint cycles,
 * sorted in decreasing order.
 *
 * @example
 * permutationCycleType([1, 2, 0, 4, 3]) // [3, 2]
 * permutationCycleType([0, 1, 2]) // [1, 1, 1]
 * permutationCycleType([1, 0, 2]) // [2, 1]
 */
export function permutationCycleType(perm: Permutation): number[] {
  const seen = new Array<boolean>(perm.length).fill(false);
  const cycleLens: number[] = [];
  for (let i = 0; i < perm.length; i++) {
    if (seen[i]) continue;
    seen[i] = true;
    let count = 1;
    let j = perm[i];
    while (!seen[j]) {
      seen[j] = true;
      j = perm[j];
      count += 1;
    }
    cycleLens.push(count);
  }
  return cycleLens.sort((a, b) => b - a);
}

/** All permutations of {0, ..., n-1}, grouped by cycle type. */
export interface CycleClass {
  cycleType: number[];
  perms: Permutation[];
}

export function allPermutationsByCycleType(n: number): CycleClass[] {
  const byKey = new Map<string, CycleClass>();
  for (const p of allPermutations(n)) {
    const cycleType = permutationCycleType(p);
    const key = cycleType.join(',');
    let entry = byKey.get(key);
    if (!entry) {
      entry = { cycleType, perms: [] };
      byKey.set(key, entry);
    }
    entry.perms.push(p);
  }
  return [...byKey.values()];
}

/**
 * The number of permutations of the same cycle type as cycleType (the size
 * of its conjugacy class in S_n, n = sum(cycleType)).
 *
 * @example
 * permutationConjugacyClassSize([1, 1, 1, 1]) // 1
 * permutationConjugacyClassSize([2, 1, 1]) // 6
 */
export function permutationConjugacyClassSize(cycleType: number[]): number {
  const counts = new Map<number, number>();
  for (const c of cycleType) counts.set(c, (counts.get(c) ?? 0) + 1);
  let z = 1;
  for (const [cycleLen, cycleCount] of counts) {
    z *= cycleLen ** cycleCount * factorial(cycleCount);
  }
  const n = cycleType.reduce((a, b) => a + b, 0);
  return Math.round(factorial(n) / z);
}
